block-storage/src/main/scala/coop/rchain/blockstorage/dag/DagRepresentation.scala

```diff
@@ -60,42 +67,20 @@ final case class DagRepresentation(
 }
 
 object DagRepresentation {
-  def empty: DagRepresentation =
-    DagRepresentation(
-      Set.empty[BlockHash],
-      Map.empty[Validator, BlockHash],
-      Map.empty[BlockHash, Set[BlockHash]],
-      SortedMap.empty[Long, Set[BlockHash]],
-      Set.empty[BlockHash],
-      none[BlockHash],
-      Set.empty[BlockHash],
-      DagMessageState[BlockHash, Validator](ByteString.EMPTY, Set(), Map())
-    )
-
   def lastFinalizedBlockUnsafe[F[_]: Sync](dr: DagRepresentation): F[BlockHash] = {
     val errMsg =
       "DagState does not contain lastFinalizedBlock. Are you calling this on empty BlockDagStorage? Otherwise there is a bug."
     dr.lastFinalizedBlockHash.liftTo[F](new Exception(errMsg))
   }
 
-  def invalidBlocks[F[_]: Sync: BlockDagStorage](dr: DagRepresentation): F[Set[BlockMetadata]] =
-    dr.invalidBlocksSet.toList.traverse(BlockDagStorage[F].lookupUnsafe).map(_.toSet)
-
   def latestMessageHash[F[_]: Sync: BlockDagStorage](
       dr: DagRepresentation,
       validator: Validator
   ): F[Option[BlockHash]] =
-    dr.latestMessagesHashes.get(validator).pure
+    dr.dagMessageState.latestMsgs.filter(_.sender == validator).map(_.id).headOption.pure
```

Comment:
With invalid blocks, validator can have multiple latest messages. We need to see how to manage this situation.

Reply:
I agree - if you remember I was saying we need latest messages be Map[Sender, Set[Message]]. And we can actually merge multiple latest messages. We can punish for that, but we still can merge.

Reply:
Here I'm thinking about places where latest message is not used like BlockApi where external code can depend on this change.
https://github.com/rchain/rchain/blob/d9cd6dde980d22dff9a96c581dbe5d7d5d7b18a2/casper/src/main/scala/coop/rchain/casper/api/BlockApiImpl.scala#L655-L661

The Map you mentioned is practically there in DAG message state.
```scala
dag.dagMessageState.latestMsgs.groupBy(_.id) : Map[BlockHash, Set[Message[BlockHash, Validator]]]
```